specs/required_block_state.md

```diff
@@ -0,0 +1,335 @@
+| MAX_BYTES_PER_NODE | uint16(32768) | - |
+| MAX_BYTES_PER_CONTRACT | uint16(32768) | - |
+| MAX_CONTRACTS_PER_BLOCK | uint16(2048) | - |
+| MAX_NODES_PER_BLOCK | uint16(32768) | - |
+| MAX_ACCOUNT_PROOFS_PER_BLOCK | uint16(8192) | - |
+| MAX_STORAGE_PROOFS_PER_ACCOUNT | uint16(8192) | - |
+
+## Definitions
+
+### `RequiredBlockState`
+
+The entire `RequiredBlockState` data format is represented by the following (SSZ-encoded and
+snappy-compressed) container.
+
+All trie nodes (account and storage) are aggregated for deduplication and simplicity.
+They are located in the `trie_nodes` members.
+A "compact" proof consists only of the root hash for that trie and the information required
+for computing the trie path. The trie nodes can then be traversed by locating the first
+node using the root hash and starting the traversal.
+
+The proof data represents values in the historical chain immediately prior to the execution of
+the block (sometimes referred to as "prestate"). That is, `RequiredBlockState` for block `n`
+contains proofs rooted in the state root of block `n - 1`.
+
+```python
+class RequiredBlockState(Container):
+    #sorted (by address)
+    compact_eip1186_proofs: List[CompactEip1186Proof, MAX_ACCOUNT_PROOFS_PER_BLOCK]
+    #sorted
+    contracts: List[Contract, MAX_CONTRACTS_PER_BLOCK]
+    #sorted
+    trie_nodes: List[TrieNode, MAX_NODES_PER_BLOCK]
+    #sorted (by block number)
+    block_hashes: List[RecentBlockHash, MAX_BLOCKHASH_READS_PER_BLOCK]
+```
+The `RequiredBlockState` is compressed using snappy encoding (see algorithms section). The
+`eth_getRequiredBlockState` JSON-RPC method returns the SSZ-encoded container with snappy encoding.
+
+### `CompactEip1186Proof`
+
+Represents the proof data whose root is the state root in the block header of the preceeding block.
+
+The account proof is obtained by calculating the account hash and traversing nodes in the
+`RequiredBlockState` container.
+
+```python
+class CompactEip1186Proof(Container):
+    address: Vector[uint8, 20]
+    balance: List[uint8, 32]
+    code_hash: Vector[uint8, 32]
+    nonce: List[uint8, 8]
+    storage_hash: Vector[uint8, 32]
+    #sorted
+    storage_proofs: List[CompactStorageProof, MAX_STORAGE_PROOFS_PER_ACCOUNT]
+```
+
+### `Contract`
+
+An alias for contract bytecode.
+```python
+Contract = List[uint8,  MAX_BYTES_PER_CONTRACT]
+```
+
+### `TrieNode`
+
+An alias for a node in a merkle patricia proof.
+
+Merkle Patricia Trie proofs consist of a list of witness nodes that correspond to each trie node that consists of various data elements depending on the type of node (e.g. blank, branch, extension, leaf).  When serialized, each witness node is represented as an RLP serialized list of the component elements.
+
+```python
+TrieNode = List[uint8,  MAX_BYTES_PER_NODE]
+```
+
+### `RecentBlockHash`
+
+A block hash accessed by the "BLOCKHASH" opcode.
+```python
+class RecentBlockHash(Container):
```

Comment:
This can be made verifiable by including the whole header here.

Reply:
Could you expand on what you mean here? 

The assumption here is that a user has a mechanism to verify the canonicality of blockhashes. 

More notes here: https://github.com/perama-v/archors#blockhash-opcode

Reply:
So my point addresses the following fault case you describe in your notes:

> A block hash is wrong: the portal node can audit all block hashes against its master accumulator prior to tracing.

If we want to solve it at the RPC level: instead of sending only block number and block hash, send the whole block header. Then the verifier can hash the header (which contains the number) and be sure about the hash.

I understand the size concern. However if the clients have to audit an external source anyway it might be worth it. Otherwise if in most use-cases the client has easy access to latest 256 blocks then we can leave out this info from `eth_getRequiredBlockState`.

Reply:
A "wrong" hash refers to a hash that is not canonical. So, the wrong hash could be a correctly computed hash of a noncanonical block. E.g. a fabricated block or an uncle block. Having the whole block included doesn't get one closer to verifying canonicality.